lib/ui/painting.dart

```diff
@@ -1695,28 +1695,42 @@ class Codec extends NativeFieldWrapperClass2 {
   void dispose() native 'Codec_dispose';
 }
 
-/// Instantiates an image codec [Codec] object.
+/// Instantiates an image [Codec].
 ///
-/// [list] is the binary image data (e.g a PNG or GIF binary data).
+/// The `list` parameter is the binary image data (e.g a PNG or GIF binary data).
 /// The data can be for either static or animated images. The following image
 /// formats are supported: {@macro flutter.dart:ui.imageFormats}
 ///
-/// The [targetWidth] and [targetHeight] arguments specify the size of the output
-/// image, in image pixels. If they are not equal to the intrinsic dimensions of the
-/// image, then the image will be scaled after being decoded. If only one dimension
-/// is specified, the omitted dimension will be scaled to maintain the original
-/// aspect ratio. If both are not specified, then the image maintains its real
-/// size.
+/// The `targetWidth` and `targetHeight` arguments specify the size of the
+/// output image, in image pixels. If they are not equal to the intrinsic
+/// dimensions of the image, then the image will be scaled after being decoded.
+/// If only one dimension is specified, the omitted dimension will be scaled to
+/// maintain the original aspect ratio. If both are not specified, then the
+/// image maintains its intrinsic size.
+///
+/// Scaling the image to larger than its intrinsic size should usually be
+/// avoided, since it causes the image to use more memory than necessary.
+/// Instead, prefer scaling the [Canvas] transform. If the image must be scaled
+/// up, the `allowUpscaling` parameter must be set to true.
 ///
 /// The returned future can complete with an error if the image decoding has
 /// failed.
-Future<Codec> instantiateImageCodec(Uint8List list, {
+Future<Codec> instantiateImageCodec(
+  Uint8List list, {
   int? targetWidth,
   int? targetHeight,
+  bool allowUpscaling = false,
```

Comment:
This is going to be a breaking change. Not sure we want to do the dance of keeping it true and the flip it later. Your call.

Reply:
For the Flutter compatibility policy, we would have to break a contributed test (including internal). I did not do a full tgp but I believe this is not breaking by that definition. I understand that is a change in default behavior though. 

I suppose to prevent any need to roll back I can default it to true for now, and then we can go back and default it to false once we can test it internally a bit more easily.

Reply:
This is a new API, so there can't be a test that breaks right? Just felt it might be in violation of the "spirit" of the compatibility policy if not the letter. Either way, your call. It was just an observation about the defaults changing.

Reply:
There could be a test somewhere that's using the targetWidth/Height to upscale the image currently, and that test would start failing. If that test is registered in flutter/tests or in google3, it counts as a breaking change.

I'm going to see if I can run this through a TGP and make sure it doesn't break anything. If not, I'm going to avoid the dance of changing defaults around - this is clearly what most users want.

Reply:
Sounds good.